**Computer Project 1** </br>

**ECE5101** 

Spencer Albano 



`CP1_sda18001.ipynb`
*can be viewed and interacted with via* 

<https://nbviewer.org/github/salbano108/ECE5101_CP1/tree/master/> 

*or by installing Jupyter Lab*
</br>
</br>

# Problem 1

For each matrix below find:

(i) The dominant eigenvalue <br />
(ii) A corresponding eigenvector <br />
(iii) The power 𝑘 used to obtain them. <br />


$$ A_{1}=
\begin{bmatrix}
  0.2 & 1.2 & 1.1 & 0.9 & 0.1 & 0 & 0.2\\ 
  0.7 & 0 & 0 & 0 & 0 & 0 & 0 \\
  0 & 0.82 & 0 & 0 & 0 & 0 & 0\\
  0 & 0 & 0.97 & 0 & 0 & 0 & 0\\
  0 & 0 & 0 & 0.97 & 0 & 0 & 0\\
  0 & 0 & 0 & 0 & 0.9 & 0 & 0\\
  0 & 0 & 0 & 0 & 0 & 0.87 & 0.2\\
\end{bmatrix}$$

$$ A_{2}=
\begin{bmatrix}
  0.25 & 0 & 0.5 & 0 & 0.5\\ 
  0 & 0 & 1 & 0 & 0\\  
  0.25 & 0.25 & 0 & 0.25 & 0.25\\ 
  0 & 0 & 0.5 & 0 & 0.5\\
  0 & 0 & 0 & 0 & 1\\ 
\end{bmatrix}$$


$$ A_{3}=
\begin{bmatrix}
  1 & 2 & 1 & 3\\ 
  2 & 1 & 2 & 3\\ 
  0 & 1 & 1 & 1\\ 
  1 & 1 & 3 & 1\\ 
\end{bmatrix}$$

---
## Solution:

In [36]:
import numpy as np

def power_method(A, max_iterations=15, precision=1e-3):
    n, m = A.shape #Find row,col of A
    x = np.ones(n) # Set the initial vector to a vector of ones
    eigenvalue_temp = 0
    
    for k in range(1, max_iterations + 1): #test from 1 to max_int to find Eigenvalue that fits the selected value of precision
        
        consant0 = x[0] # Store first component of previous vector
        x = np.dot(A, x) # Multiply by matrix A to get the next vector
        consant1 = x[0] # Store first component of current vector


        x = x / x[0] # Divide each vector by their first component to scale

        eigenvalue = consant1/consant0 # Approximate the eigenvalue from (k+1)/(k) vector results

        # Check if eigenvalue is still changing for selected level of precision
        if abs(eigenvalue - eigenvalue_temp) < precision:
            eigenvalue = np.round(eigenvalue, 3)
            eigenvector = np.round(x, 3)
            return eigenvalue, eigenvector, k

        eigenvalue_temp = eigenvalue # Update the temporary eigenvalue for the next iteration

### Example Matrix $$A = \begin{bmatrix}  1 & 2 & 0 \\   2 & 1 & 2 \\   1 & 2 & 3 \\  \end{bmatrix}$$

In [35]:
A = np.array([[1, 2, 0], [2, 1, 2], [1, 2, 3]])
eigenvalue, eigenvector, k = power_method(A)
np.set_printoptions(precision=3, floatmode='fixed')
print("Dominant Eigenvalue:", eigenvalue)
print("Eigenvector:", eigenvector)
print("K:", k)

Dominant Eigenvalue: 4.82
Eigenvector: [1.000 1.910 2.648]
K: 9


$$ \lambda_{1}v_{1} \approx 4.82 \begin{bmatrix}  1.000 \\   1.910 \\   2.648 \\  \end{bmatrix} for \: k=9 $$ 


---
### Matrix 1 $$ A_{1}=\begin{bmatrix} 0.2 & 1.2 & 1.1 & 0.9 & 0.1 & 0 & 0.2\\   0.7 & 0 & 0 & 0 & 0 & 0 & 0 \\  0 & 0.82 & 0 & 0 & 0 & 0 & 0\\  0 & 0 & 0.97 & 0 & 0 & 0 & 0\\  0 & 0 & 0 & 0.97 & 0 & 0 & 0\\  0 & 0 & 0 & 0 & 0.9 & 0 & 0\\  0 & 0 & 0 & 0 & 0 & 0.87 & 0.2\\ \end{bmatrix}$$

In [38]:
A1 = np.array([[0.2, 1.2, 1.1, 0.9, 0.1, 0, 0.2], [0.7, 0, 0, 0, 0, 0, 0], [0, 0.82, 0, 0, 0, 0, 0], [0, 0, 0.97, 0, 0, 0, 0],[0, 0, 0, 0.97, 0, 0, 0], [0, 0, 0, 0, 0.9, 0, 0], [0, 0, 0, 0, 0, 0.87, 0.2]])
eigenvalue, eigenvector, k = power_method(A1)
np.set_printoptions(precision=3, floatmode='fixed')
print("Dominant Eigenvalue:", eigenvalue)
print("Eigenvector:", eigenvector)
print("K:", k)

Dominant Eigenvalue: 1.372
Eigenvector: [1.000 0.510 0.305 0.213 0.156 0.098 0.074]
K: 9


$$ \lambda_{1}v_{1} \approx 1.372 \begin{bmatrix}  1.000 \\   0.510 \\   0.305 \\ 0.213 \\ 0.156 \\ 0.098 \\ 0.074 \end{bmatrix} for \: k=9 $$ 


---
### Matrix 2 $$ A_{2}= \begin{bmatrix}0.25 & 0 & 0.5 & 0 & 0.5\\   0 & 0 & 1 & 0 & 0\\    0.25 & 0.25 & 0 & 0.25 & 0.25\\   0 & 0 & 0.5 & 0 & 0.5\\  0 & 0 & 0 & 0 & 1\\ \end{bmatrix}$$




In [39]:
A2 = np.array([[0.25, 0, 0.5, 0, 0.5], [0, 0, 1, 0, 0], [0.25, 0.25, 0, 0.25, 0.25], [0, 0, 0.5, 0, 0.5],[0, 0, 0, 0, 1]])
eigenvalue, eigenvector, k = power_method(A2)
np.set_printoptions(precision=3, floatmode='fixed')
print("Dominant Eigenvalue:", eigenvalue)
print("Eigenvector:", eigenvector)
print("K:", k)

Dominant Eigenvalue: 1.009
Eigenvector: [1.000 0.802 0.807 0.752 0.702]
K: 7


$$ \lambda_{1}v_{1} \approx 1.009 \begin{bmatrix}  1.000 \\   0.802 \\   0.807 \\ 0.752 \\ 0.702 \end{bmatrix} for \: k=7 $$ 


---
### Matrix 3 $$ A_{3}= \begin{bmatrix}  1 & 2 & 1 & 3\\   2 & 1 & 2 & 3\\   0 & 1 & 1 & 1\\   1 & 1 & 3 & 1\\ \end{bmatrix}$$

In [39]:
A2 = np.array([[0.25, 0, 0.5, 0, 0.5], [0, 0, 1, 0, 0], [0.25, 0.25, 0, 0.25, 0.25], [0, 0, 0.5, 0, 0.5],[0, 0, 0, 0, 1]])
eigenvalue, eigenvector, k = power_method(A2)
np.set_printoptions(precision=3, floatmode='fixed')
print("Dominant Eigenvalue:", eigenvalue)
print("Eigenvector:", eigenvector)
print("K:", k)

Dominant Eigenvalue: 1.009
Eigenvector: [1.000 0.802 0.807 0.752 0.702]
K: 7
